# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
# Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
# Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography','Gender',	'Age',	'Tenure','Balance',	'NumOfProducts',	'HasCrCard',	'IsActiveMember',	'EstimatedSalary',	'Exited']].copy()

## Preprocessing

In [ ]:
df = pd.DataFrame(df)

# Memeriksa tipe data kolom-kolom
categorical_columns = df.select_dtypes(include=['object']).columns
categorical_columns

Index(['Geography', 'Gender'], dtype='object')

In [ ]:
# Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies
df_encoded = pd.get_dummies(df, columns=categorical_columns)

# Menampilkan DataFrame yang telah diencode
df_encoded.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df_encoded.drop(columns=['Exited'])  # Memisahkan semua kolom kecuali "Exited" sebagai fitur
y = df_encoded['Exited']  # Menggunakan kolom "Exited" sebagai target

In [ ]:
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1


Model yang digunakan dalam contoh kode yang Anda berikan adalah **Logistic Regression**

dilakukan tuning parameter menggunakan GridSearchCV untuk mencari kombinasi hyperparameter terbaik dari model tersebut yang didefinisikan dalam param_grid, dan mencari yang terbaik berdasarkan metrik 'accuracy' dengan validasi silang (cross-validation) menggunakan 10 fold cross-validation. Hasil terbaik dari pencarian parameter akan diberikan oleh grid.best_params_.







In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh
model1 = LogisticRegression()
params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # cross validation
            )

grid.fit(X_train,y_train)
grid.best_params_

{'C': 0.5, 'tol': 0.1}

In [ ]:
# Evaluasi Model
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1929   54]
 [ 412  105]]

0.8136


Model pertama dengan menggunakan Logistic Regression menghasilkan akurasi sebesar 81,36%

## Model2
### Soal :
Model yang digunakan adalah **Random Forest Classifier**  yang menggabungkan sejumlah besar pohon keputusan yang dikenal sebagai "pohon acak" (random trees).

jumlah estimator (n_estimators) : 100, 200, 300

kedalaman maksimum (max_depth) :  [None, 10, 20, 30]


min_samples_split : [2, 5, 10]


min_samples_leaf : [1,2,4]

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model Random Forest Classifier
model2 = RandomForestClassifier()

# Definisikan daftar hyperparameter yang ingin Anda tuning
param_grid = {'n_estimators': [100, 200, 300],
              'max_depth': [None, 10, 20, 30],
              'min_samples_split': [2, 5, 10],
              'min_samples_leaf': [1, 2, 4]}

# Inisialisasi objek GridSearchCV
grid_search = GridSearchCV(estimator=model2, param_grid=param_grid,
                           scoring='accuracy', cv=10, n_jobs=-1)

# Melakukan pencarian hyperparameter terbaik
grid_search.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik
best_params = grid_search.best_params_
print("Hyperparameter terbaik:", best_params)

# Model Random Forest Classifier dengan hyperparameter terbaik
best_model2 = grid_search.best_estimator_

# Melatih model Random Forest Classifier dengan hyperparameter terbaik
best_model2.fit(X_train, y_train)

Hyperparameter terbaik: {'max_depth': 20, 'min_samples_leaf': 4, 'min_samples_split': 2, 'n_estimators': 300}


RandomForestClassifier(max_depth=20, min_samples_leaf=4, n_estimators=300)

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = best_model2.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.45      0.57       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.85      0.86      0.85      2500


[[1919   64]
 [ 283  234]]

0.8612


Akurasi yang diperoleh dengan model Random Forest Classifier adalah sebesar 86,12%

## Model3
Model ketiga yang digunakan adalah **Gradient Boosting Classifier**  

jumlah estimator (n_estimators) : [50, 100,200]

learning rate : [0.01, 0.1, 0.2]

max depth : [3, 4, 5]

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

# Inisialisasi model Gradient Boosting Classifier
model3 = GradientBoostingClassifier()

# Definisikan daftar hyperparameter yang ingin Anda tuning
params = {'n_estimators': [50, 100, 200],
          'learning_rate': [0.01, 0.1, 0.2],
          'max_depth': [3, 4, 5]}

# Inisialisasi objek GridSearchCV
grid_search = GridSearchCV(estimator=model3, param_grid=params,
                           scoring='accuracy', cv=10, n_jobs=-1)

# Melakukan pencarian hyperparameter terbaik
grid_search.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik
best_params = grid_search.best_params_
print("Hyperparameter terbaik:", best_params)

# Melatih model Gradient Boosting Classifier dengan hyperparameter terbaik
best_model3 = grid_search.best_estimator_
best_model3.fit(X_train, y_train)

Hyperparameter terbaik: {'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 100}


GradientBoostingClassifier(max_depth=4)

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = best_model3.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.78      0.51      0.62       517

    accuracy                           0.87      2500
   macro avg       0.83      0.74      0.77      2500
weighted avg       0.86      0.87      0.86      2500


[[1907   76]
 [ 251  266]]

0.8692


Menggunakan model Gradient Boosting Clasifier diperoleh akurasi sebeasar 86,92%

## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Dari percobaan ketiga model di atas terlihat bahwa algoritma Gradient Boosting memiliki akurasi paling tinggi yaitu sebesar  86.92%., sehingga untuk memprediksi pelanggan bank yang nonaktif dapat menggunakan model algoritmaa Gradient Boosting